### DISCUSSION TITLE
Can the m.Freeze and m.Thawed be canceled?

### DISCUSSION THREAD

**Discussion by lijinyong** [opencontainers/runc#3914] [General]
The lack of error handling for m.Freeze and m.Thawed in the code may cause errors when freezing and thawing cgroups, which may affect the correctness of the program.

For example, when executing m.Freeze, if runc exits abnormally (due to fault injection or other abnormal scenarios), all business processes within the cgroup will be frozen.

// signalAllProcesses freezes then iterates over all the processes inside the
// manager's cgroups sending the signal s to them.
// If s is SIGKILL then it will wait for each process to exit.
// For all other signals it will check if the process is ready to report its
// exit status and only if it is will a wait be performed.
func signalAllProcesses(m cgroups.Manager, s os.Signal) error {
	var procs []*os.Process
	if err := m.Freeze(configs.Frozen); err != nil {
		logrus.Warn(err)
	}
	pids, err := m.GetAllPids()
	if err != nil {
		if err := m.Freeze(configs.Thawed); err != nil {
			logrus.Warn(err)
		}
		return err
	}
	for _, pid := range pids {
		p, err := os.FindProcess(pid)
		if err != nil {
			logrus.Warn(err)
			continue
		}
		procs = append(procs, p)
		if err := p.Signal(s); err != nil {
			logrus.Warn(err)
		}
	}
	if err := m.Freeze(configs.Thawed); err != nil {
		logrus.Warn(err)
	}

**Comment by lifubang** [comment]
I think they can't be canceled, one obvious reason is that it may cause `PID` reuse attack.
As you say, If `runc` exits abnormally, we can call `runc` again to retry. Because `signalAllProcesses` is used in a separate `RUNC` command process. The command caller should have detect and retry mechanism.

**Reply by lijinyong** [reply]
After the container is frozen,  runc is not allowed to be executed  in a paused container. and all retry repair mechanisms that ultimately rely on runc will of course fail. 

Although runc provides a new feature ignore-paused to support the execution of runc in the pause container, containedred seems to have no intention of integration.

**Reply by lifubang** [reply]
But `runc kill` is allowed to be executed for a paused container.
